### PR TITLE
Rewrite metric generation to use fewer DB queries.

### PR DESCRIPTION
**Pull request by progala**
- Rewrite metric generation to use fewer DB queries
- Make metric names more explicit
- Ensure only one metric is generated for inventory items with the same part_id
- Change the HW out of support logic. Only HW notices with out of support dates before today's date are considered to be out of support.

### CODE REVIEW

File: nautobot_device_lifecycle_mgmt/metrics.py [hunk 1]
```diff
@@ -1,69 +1,123 @@
...
+    )
 
-    part_number_gauge = GaugeMetricFamily(
-        "nautobot_lcm_devices_eos_per_part_number", "Nautobot LCM Devices EOS per Part Number", labels=["part_number"]
+    today = datetime.today().date()
+    hw_end_of_support = HardwareLCM.objects.filter(end_of_support__lt=today)
+    hw_end_of_support_device_types = hw_end_of_support.exclude(device_type__isnull=True).values_list(
+        "device_type", flat=True
+    )
+    hw_eos_inventoryitems = hw_end_of_support.exclude(inventory_item__isnull=True).values_list(
```
Comment (pato23arg) [inline nautobot_device_lifecycle_mgmt/metrics.py]: minor detail here, shouldn't it be:
`hw_end_of_support_inventoryitems` ?

Comment (progala) [reply nautobot_device_lifecycle_mgmt/metrics.py]: Good catch! I'll get it updated.

**Comment by nkallergis** [comment]
Two small notes on an otherwise fine PR:
1. If I'm not missing something, in order to create a Hardware Notice for an Inventory Number, it is required that it has a Part Number. Therefore, I would be tempted to not handle at all the cases where a PN is missing (lines 72-80). No strong opinion on it though.
2. I haven't benchmarked the two approaches but I'm fairly sure that the django-based one will be more performant. On the other hand, judging from my personal level of Django knowledge, it's a bit harder to understand/maintain (`annotate`, `Coalesce` aren't even in my English vocabulary 😆). The question is: are the performance gains worth the readability loss? I guess the answer is yes but I'd love to hear other people's opinions too!

Overall, LGTM!

**Comment by pato23arg** [comment]
> Two small notes on an otherwise fine PR:
> 
>     1. If I'm not missing something, in order to create a Hardware Notice for an Inventory Number, it is required that it has a Part Number. Therefore, I would be tempted to not handle at all the cases where a PN is missing (lines 72-80). No strong opinion on it though.
> 
>     2. I haven't benchmarked the two approaches but I'm fairly sure that the django-based one will be more performant. On the other hand, judging from my personal level of Django knowledge, it's a bit harder to understand/maintain (`annotate`, `Coalesce` aren't even in my English vocabulary 😆). The question is: are the performance gains worth the readability loss? I guess the answer is yes but I'd love to hear other people's opinions too!
> 
> 
> Overall, LGTM!

for 1. I think unfortunately inventory items allow empty part ids. i.e. an inventory item that has no notice but exists in nautobot database.
for 2. @progala measured it and even in dev performance gains are > 50% with regards to query RTT!

**Comment by nkallergis** [comment]
Approved! :)

As for the question about inventory items without a part number, personally I wouldn't add them but it's obviously 100% a decision that you people have to make.

**Review by nkallergis — APPROVED** [review]

**Review by pato23arg — APPROVED** [review]

**Pull request merged**